File: problems/1301-1400/1396_design-underground-system/solution.js
class UndergroundSystem {
    constructor() {
        this.checkins = new Map();
        this.totals = new Map();
    }

    checkIn(id, stop, t) {
        this.checkins.set(id, { stop, t });
    }

    checkOut(id, stop, t) {
        const { stop: start, t: started } = this.checkins.get(id);
        this.checkins.delete(id);
        let byEnd = this.totals.get(start);
        if (!byEnd) {
            byEnd = new Map();
            this.totals.set(start, byEnd);
        }
        let bucket = byEnd.get(stop);
        if (!bucket) {
            bucket = [0, 0];
            byEnd.set(stop, bucket);
        }
        bucket[0] += t - started;
        bucket[1] += 1;
    }

    getAverageTime(fromStop, toStop) {
        const [total, count] = this.totals.get(fromStop).get(toStop);
        return total / count;
    }
}

File: problems/1301-1400/1396_design-underground-system/solution.rs
use std::collections::HashMap;

pub struct UndergroundSystem {
    checkins: HashMap<i32, (String, i32)>,
    totals: HashMap<(String, String), (i64, i64)>,
}

impl UndergroundSystem {
    pub fn new() -> Self {
        UndergroundSystem {
            checkins: HashMap::new(),
            totals: HashMap::new(),
        }
    }

    pub fn checkIn(&mut self, id: i32, stop: String, t: i32) {
        self.checkins.insert(id, (stop, t));
    }

    pub fn checkOut(&mut self, id: i32, stop: String, t: i32) {
        let (start, started) = self.checkins.remove(&id).unwrap();
        let bucket = self.totals.entry((start, stop)).or_insert((0, 0));
        bucket.0 += (t - started) as i64;
        bucket.1 += 1;
    }

    pub fn getAverageTime(&mut self, fromStop: String, toStop: String) -> f64 {
        let &(total, count) = self.totals.get(&(fromStop, toStop)).unwrap();
        total as f64 / count as f64
    }
}
